sql: re-vamp the view dependency analysis

knz opened the pull request:
Prior to this patch, view dependency analysis during CREATE VIEW was broken because it would only check dependencies after plan optimization, i.e. possibly after some dependency were lost. For example, with the queries:  

```sql
CREATE VIEW v AS SELECT k FROM (SELECT k,v FROM kv) -- loses dependency on kv.v 
CREATE VIEW v AS SELECT k,v FROM kv WHERE FALSE -- loses dependency on kv
 ```  

This patch addresses the issue as follows:  
- the dependencies are now collected during the initial construction   of the query plan, before any optimization are applied. This way we  ensure the dependency tracking is complete. 
- a migration is implemented that will fix any view descriptor and   corresponding dependency information that was populated prior to this fix.

Fixes #17306.
Fixes #17290.
Replaces 74c7b0fd20d3da721c0f2323fc09be480d3244ab (reverted via #17305)
Needed for #15388.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17310)
<!-- Reviewable:end -->

knz commented:
okay so:
- @a-robinson @benesch please have a look at my migration test and see if that is well aligned with the spirit of the thing
- @jordanlewis @andreimatei @RaduBerinde one of you please have a look at the new dependency analysis code and provide feedback

a-robinson commented:
I'm not sure dropping the view's dependencies and checking that the views don't work without them was really needed, but the test looks reasonable to me.

---

Reviewed 2 of 33 files at r1.
Review status: 2 of 33 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-:-KqO8OyQyiGcgV5zLyZ5:b-d42lwq)*
<!-- Sent from Reviewable.io -->

benesch commented on pkg/migrations/migrations_test.go:
```diff
@@ -502,3 +505,217 @@ func TestCreateSystemTable(t *testing.T) {
+	defer s.Stopper().Stop(ctx)
+	e := s.Executor().(*sql.Executor)
+
+	t.Run("create test tables", func(t *testing.T) {
```
It's not clear to me that these subtests are providing any value, since they can't be run independently of one another. Plus, each subtest creates its own session, which adds quite a bit of boilerplate. @knz, does this exercise some codepath that wouldn't be executed if you used the same session throughout?

benesch commented on pkg/migrations/migrations_test.go:
```diff
@@ -502,3 +505,217 @@ func TestCreateSystemTable(t *testing.T) {
+	})
+
+	// Now, corrupt the descriptors by breaking their dependency information.
+	t.Run("break descriptors", func(t *testing.T) {
```
E.g., this will segfault if someone runs `make test PKG=./pkg/migration TESTS=TestUpdateViewDependenciesMigration/break_descriptors`. I feel like you might be better off with `t.Logf("break descriptors")`.

knz commented:
Test simplified thanks to generous suggestions by Jordan and Nikhil.

---

Review status: 1 of 29 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/migrations/migrations_test.go, line 539 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqNqtUz9Ua_Eq0Ve_uG-r1-539:-KqSxnuzCUyG4DEpwFHD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/646c05d14c9a110522c2d08afef18972c85598d9/pkg/migrations/migrations_test.go#L539)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

It's not clear to me that these subtests are providing any value, since they can't be run independently of one another. Plus, each subtest creates its own session, which adds quite a bit of boilerplate. @knz, does this exercise some codepath that wouldn't be executed if you used the same session throughout?
</blockquote></details>

Done.

---

*[pkg/migrations/migrations_test.go, line 598 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqNqtUz9Ua_Eq0Ve_uG-r1-598:-KqSxoo-5NfJuND-6-kp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/646c05d14c9a110522c2d08afef18972c85598d9/pkg/migrations/migrations_test.go#L598)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

E.g., this will segfault if someone runs `make test PKG=./pkg/migration TESTS=TestUpdateViewDependenciesMigration/break_descriptors`. I feel like you might be better off with `t.Logf("break descriptors")`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-:-KqTO79MmcFgSzgAYNEL:b-h41kwt)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM

The second "fixes" in the PR comment links to a PR. Is that what you wanted?


After we talked, I thought some more: for the future, I still think that having a plan construction mode where all optimizations are disabled, and then collecting the dependencies by visiting that, seems like a cleaner way than your proposed followed: the recording `SchemaAccessor`. You said that it wouldn't work because the optimizations change the typing of the view but a) I don't see why that's true and b) you don't have to do anything with the unoptimized plan besides collect the dependencies; you can construct a new, optimized plan afterwards for typing and whatever else.
One problem is see is the index dependencies (that's a thing, right?) if the index selection changes with optimizations. But perhaps we can build both the optimized and the unoptimized plans and merge the dependencies from both?
Does this make sense? I'm thinking that a plan creation mode without any optimizations could be more generally useful.

---

Review status: 0 of 19 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/create.go, line 334 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTflo17YbOKJB9wB-2:-KqTflo17YbOKJB9wB-3:b-vobh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L334)):*
> ```Go
> // table -- which index and columns are being depended upon.
> type planDependencyInfo struct {
> 	desc *sqlbase.TableDescriptor
> ```

why is this a pointer?

---

*[pkg/sql/create.go, line 382 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTgpqDQIbm66MI4cfa:-KqTgpqDQIbm66MI4cfb:b-q2w50w) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L382)):*
> ```Go
> 	// The traversal will update the NormalizableTableNames in-place,
> 	// so the changes are persisted in n.AsSource. We discard
> 	// the result of pretty-printing.
> ```

I was confused about what the reference to pretty-printing is telling me.
I'd say something like "we use FormatNode` as a tree visitor".

---

*[pkg/sql/create.go, line 398 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqThLZctl3lyUveJ-Dk:-KqThLZctl3lyUveJ-Dl:bmt0xfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L398)):*
> ```Go
> 				}
> 				// Persist the database prefix expansion.
> 				tn.DBNameOriginallyOmitted = false
> ```

nit: I don't know anything about anything, but this line suggests to me that `DBNameOriginallyOmitted` is badly named, since we're lying to it. Perhaps it should be named based on its effect, which I guess is something like "UseOriginalUnqualifiedWhenPrinting`.

---

*[pkg/sql/create.go, line 414 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqThy9o3LhRf3bE1M6v:-KqThy9o3LhRf3bE1M6w:b9lgj71) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L414)):*
> ```Go
> 
> 	// Request dependency tracking.
> 	defer func(prev planDependencies) { p.planDeps = prev }(p.planDeps)
> ```

I don't understand what this does and how it's "requesting" anything. You're restoring `p.planDeps` at the end?

---

*[pkg/sql/create.go, line 423 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTiBwev9YKyNgftc4J:-KqTiBwflxvZcIbPcGyC:bkx0uhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L423)):*
> ```Go
> 	}
> 
> 	log.Infof(ctx, "collected view dependencies:\n%s", p.planDeps.String())
> ```

`Infof` seems high to me

---

*[pkg/sql/create.go, line 423 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTiMQK7vLBQ_mw_k71:-KqTiMQK7vLBQ_mw_k72:b-u8h0u4) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L423)):*
> ```Go
> 	}
> 
> 	log.Infof(ctx, "collected view dependencies:\n%s", p.planDeps.String())
> ```

isn't `p.planDeps` empty here? So confused.

---

*[pkg/sql/create.go, line 464 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTiYqG-Fkf0uMfn_8T:-KqTiYqG-Fkf0uMfn_8U:b2q0uox) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L464)):*
> ```Go
> // view descriptors created prior to fixing #17269 and #17306;
> // it may also be used by a future "fsck" utility.
> func RecomputeViewDependencies(ctx context.Context, txn *client.Txn, e *Executor) error {
> ```

bleah. I think we should move this function to another file that nobody needs to see, specific to migrations.

---

*[pkg/sql/create.go, line 669 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTwUxynZbc511GWjzs:-KqTwUxynZbc511GWjzt:b-9tnx7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L669)):*
> ```Go
> 		for _, dep := range updated.deps {
> 			// The logical plan constructor merely registered the dependencies.
> 			// It did not populate the "ID" field of TableDescriptor_Reference.
> ```

why didn't the ctor populate this field? Is it didn't know its own id?

---

*[pkg/sql/planner.go, line 77 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqU19GcbLGR6m3a79T_:-KqU19GcbLGR6m3a79Ta:b7r1cqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/planner.go#L77)):*
> ```Go
> 	cancelChecker CancelChecker
> 
> 	// planDeps, if non-nil, collects the table/view dependencies for this query.
> ```

please expand the comment saying something about the responsibilities of the node ctors when this is non-nil: to record what descriptors they're using.

And then get rid of this member soon :P

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-:-KqTWVb_OIgG4-0Yw56O:b7ypmbh)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 19 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/create.go, line 334 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTflo17YbOKJB9wB-2:-KqUwlJ9BCdqb2szp6oO:b-tltmvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L334)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why is this a pointer?
</blockquote></details>

Added a clarifying comment.

---

*[pkg/sql/create.go, line 382 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTgpqDQIbm66MI4cfa:-KqUy2fDHIiQ6a7hMYyY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L382)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I was confused about what the reference to pretty-printing is telling me.
I'd say something like "we use FormatNode` as a tree visitor".
</blockquote></details>

Done.

---

*[pkg/sql/create.go, line 398 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqThLZctl3lyUveJ-Dk:-KqUy4l4o3J55TkIS_6J:b-ohhihr) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L398)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: I don't know anything about anything, but this line suggests to me that `DBNameOriginallyOmitted` is badly named, since we're lying to it. Perhaps it should be named based on its effect, which I guess is something like "UseOriginalUnqualifiedWhenPrinting`.
</blockquote></details>

I agree. Filed this separately as #17374.

---

*[pkg/sql/create.go, line 414 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqThy9o3LhRf3bE1M6v:-KqUySk-6yzLXL_2Sms9:bll3e6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L414)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't understand what this does and how it's "requesting" anything. You're restoring `p.planDeps` at the end?
</blockquote></details>

Yes! This is a general pattern since this planNode constructor is really a visitor, and there may be a parent with a different idea of what planDeps should be. We do the same with `avoidCachedDescriptors` above for the same reason.

---

*[pkg/sql/create.go, line 423 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTiMQK7vLBQ_mw_k71:-KqUyi7TAfQjD458BWPb:bao66a9) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L423)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

isn't `p.planDeps` empty here? So confused.
</blockquote></details>

Read the code above :)

---

*[pkg/sql/create.go, line 423 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTiBwev9YKyNgftc4J:-KqUyjMg6pI3fqI431Tj:br7bwct) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L423)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`Infof` seems high to me
</blockquote></details>

Indeed. Changed to `VEventf`.

---

*[pkg/sql/create.go, line 464 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTiYqG-Fkf0uMfn_8T:-KqUynG-L6I7vkGdPENg:b18m5k8) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L464)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

bleah. I think we should move this function to another file that nobody needs to see, specific to migrations.
</blockquote></details>

I am going to refactor this code in the next review iteration to put all the view-related stuff in its own file. How does that sound?

---

*[pkg/sql/create.go, line 669 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTwUxynZbc511GWjzs:-KqUyuSHHlB_TEsiBBSC:b-6z5r77) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L669)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why didn't the ctor populate this field? Is it didn't know its own id?
</blockquote></details>

Explained in the comment above the definition of the field.

---

*[pkg/sql/planner.go, line 77 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqU19GcbLGR6m3a79T_:-KqUz9aCua6rCUaPuY6s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/planner.go#L77)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please expand the comment saying something about the responsibilities of the node ctors when this is non-nil: to record what descriptors they're using.

And then get rid of this member soon :P
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 23 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/sql/create.go, line 464 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTiYqG-Fkf0uMfn_8T:-KqV5a94RWIFfTI7Rlie:b-gqb7u3) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L464)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I am going to refactor this code in the next review iteration to put all the view-related stuff in its own file. How does that sound?
</blockquote></details>

Ok done already.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Reviewed 1 of 33 files at r2.
Review status: 1 of 23 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/migration/sqlmigrations/migrations.go, line 420 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqVDGrDN3HbSr8kO3qc:-KqVDGrDN3HbSr8kO3qd:bufet3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9affb59e067cc84b747976da9416cec9692f5e0/pkg/migration/sqlmigrations/migrations.go#L420)):*
> ```Go
> 
> // repopulateViewDeps ensures that each view V that depends on a table
> // T depends on all columns in T. (#17269 #17306)
> ```

this comment talks about _all columns_ but `sql.RecomputeViewDeps` doesn't (so I assume it just computes the correct cols). Which one is it? Or what does "all" mean?

---

*[pkg/migration/sqlmigrations/migrations.go, line 422 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqVCHrxWd3kafa5wZyu:-KqVCHrxWd3kafa5wZyv:b-dnohwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9affb59e067cc84b747976da9416cec9692f5e0/pkg/migration/sqlmigrations/migrations.go#L422)):*
> ```Go
> // T depends on all columns in T. (#17269 #17306)
> func repopulateViewDeps(ctx context.Context, r runner) error {
> 	// System tables can only be modified by a privileged internal user.
> ```

I don't know what this comment about the privileged user is telling me. Is it about this code, or an assumption about the caller? I don't see any use anywhere around here.

---

*[pkg/migration/sqlmigrations/migrations.go, line 423 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqVD_8MOOoM1Dfhid_H:-KqVD_8ND-a0Wejvrozv:bkateiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9affb59e067cc84b747976da9416cec9692f5e0/pkg/migration/sqlmigrations/migrations.go#L423)):*
> ```Go
> func repopulateViewDeps(ctx context.Context, r runner) error {
> 	// System tables can only be modified by a privileged internal user.
> 	// Retry a limited number of times because returning an error and letting
> ```

is this comment about retrying necessary (and talking about panicking at the wrong level - this code doesn't panic)? Nobody would be tempted to try something forever (particularly since this interface supports returning an error), would they?

---

*[pkg/sql/create.go, line 414 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqThy9o3LhRf3bE1M6v:-KqVFTQE_p-Rbil_NZ-1:buw0oul) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L414)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes! This is a general pattern since this planNode constructor is really a visitor, and there may be a parent with a different idea of what planDeps should be. We do the same with `avoidCachedDescriptors` above for the same reason.
</blockquote></details>

ummm isn't this horrible? If there were a parent which cares about planDeps, wouldn't it also care about the deps accumulated under this view? I don't know the answer to that question, so:
Can there really be a parent on top of a `CREATE VIEW`? If not, I'd suggest we find a way to assert that there is no such parent through the planner, and removing this code.

---

*[pkg/sql/create.go, line 669 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTwUxynZbc511GWjzs:-KqVIWRz-FkkKOYKXK_R:b-y06etv) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L669)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Explained in the comment above the definition of the field.
</blockquote></details>

which field?

---

*[pkg/sql/views.go, line 35 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqVEGFJ41JmPIWw8_kp:-KqVEGFJ41JmPIWw8_kq:b-jh08bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9affb59e067cc84b747976da9416cec9692f5e0/pkg/sql/views.go#L35)):*
> ```Go
> 	// desc is a reference to the descriptor for the table being
> 	// depended on. We keep a reference and not a copy because
> 	// dependency analysis does not need to modify the descriptor.
> ```

well but what about if someone else wants to modify the descriptor. If there's no good reason, save me from thinking about anything and make a copy of the proto. This is just used for schema operations, right? So performance doesn't matter.

---

*[pkg/sql/views.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqVMwfOgc7OFOEDvlVp:-KqVMwfOgc7OFOEDvlVq:b-6u52ov) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9affb59e067cc84b747976da9416cec9692f5e0/pkg/sql/views.go#L79)):*
> ```Go
> ) (planDependencies, sqlbase.ResultColumns, error) {
> 	// To avoid races with ongoing schema changes to tables that the view
> 	// depends on, make sure we use the most recent versions of table
> ```

I spent a bunch of time being confused around here; I think this comment misses the point: what's important is not (only) that we use the current version of descriptors, but that the reads and writes of descriptors are transactional. The comment should reference `saveNonmutationAndNotify` which does the write.

Also, I think that function needs a comment about how the transaction would better have read the descriptor that it's writing. And to make it safe, I'd also change it to be a CPut and test the existing value. I'd also comment on the definition of `avoidCachedDescriptors` to highlight this use (which now might be the only use, after @vivekmenezes makes all caches timestamp-aware): the fact that the current txn is going to write descriptors and so it needs to make sure it actually reads them first.

We should also assert here that the transaction is SERIALIZABLE and refuse to allow creation of views in SNAPSHOT transactions :S

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310)*
<!-- Sent from Reviewable.io -->

knz commented:
Thank you for your insightful comment, especially the very valuable insight on avoidCachedDescriptors. I will investigate this further, I think we may have some more bugs lurking.

---

Review status: 0 of 23 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/migration/sqlmigrations/migrations.go, line 420 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqVDGrDN3HbSr8kO3qc:-KqX7E2k8xcQjjPXX5q7:bao7y5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9affb59e067cc84b747976da9416cec9692f5e0/pkg/migration/sqlmigrations/migrations.go#L420)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this comment talks about _all columns_ but `sql.RecomputeViewDeps` doesn't (so I assume it just computes the correct cols). Which one is it? Or what does "all" mean?
</blockquote></details>

Moved the comment to RecomputeViewDependencies where it belongs.

---

*[pkg/migration/sqlmigrations/migrations.go, line 422 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqVCHrxWd3kafa5wZyu:-KqX7BwUb-M5rRgHELRm:b-5kh9qv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9affb59e067cc84b747976da9416cec9692f5e0/pkg/migration/sqlmigrations/migrations.go#L422)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't know what this comment about the privileged user is telling me. Is it about this code, or an assumption about the caller? I don't see any use anywhere around here.
</blockquote></details>

The comment was stale, removed it. Thanks for noticing.

---

*[pkg/migration/sqlmigrations/migrations.go, line 423 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqVD_8MOOoM1Dfhid_H:-KqX7dpymrXo8sHwaCcA:b-bq3v22) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9affb59e067cc84b747976da9416cec9692f5e0/pkg/migration/sqlmigrations/migrations.go#L423)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is this comment about retrying necessary (and talking about panicking at the wrong level - this code doesn't panic)? Nobody would be tempted to try something forever (particularly since this interface supports returning an error), would they?
</blockquote></details>

Yes you're right. I simplified the code instead.

---

*[pkg/sql/create.go, line 414 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqThy9o3LhRf3bE1M6v:-KqX41vlJXdkEuNfiOVI:b-7txd8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L414)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ummm isn't this horrible? If there were a parent which cares about planDeps, wouldn't it also care about the deps accumulated under this view? I don't know the answer to that question, so:
Can there really be a parent on top of a `CREATE VIEW`? If not, I'd suggest we find a way to assert that there is no such parent through the planner, and removing this code.
</blockquote></details>

I am going to respectfully and politely ignore this comment. This is really a common pattern in tree traversals. We can discuss more offline.

---

*[pkg/sql/create.go, line 669 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqTwUxynZbc511GWjzs:-KqX4HubWFpXzMX6COjg:b-8x08xc) ([raw file](https://github.com/cockroachdb/cockroach/blob/285fb223a6df5c60675095af067f340243fedfa8/pkg/sql/create.go#L669)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

which field?
</blockquote></details>

Extended this comment to clarify.

---

*[pkg/sql/views.go, line 35 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqVEGFJ41JmPIWw8_kp:-KqX4Wp3XD4i0uUXHUxn:b-clskcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9affb59e067cc84b747976da9416cec9692f5e0/pkg/sql/views.go#L35)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well but what about if someone else wants to modify the descriptor. If there's no good reason, save me from thinking about anything and make a copy of the proto. This is just used for schema operations, right? So performance doesn't matter.
</blockquote></details>

Peter and I thoroughly checked together that referenced descriptors are not being modified anywhere, see the discussion on #17309.
There's absolutely no good reason to waste memory. We do the same in scanNode now.

---

*[pkg/sql/views.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqVMwfOgc7OFOEDvlVp:-KqX4hwilHeorhP6Kg95:b-p3en0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9affb59e067cc84b747976da9416cec9692f5e0/pkg/sql/views.go#L79)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I spent a bunch of time being confused around here; I think this comment misses the point: what's important is not (only) that we use the current version of descriptors, but that the reads and writes of descriptors are transactional. The comment should reference `saveNonmutationAndNotify` which does the write.

Also, I think that function needs a comment about how the transaction would better have read the descriptor that it's writing. And to make it safe, I'd also change it to be a CPut and test the existing value. I'd also comment on the definition of `avoidCachedDescriptors` to highlight this use (which now might be the only use, after @vivekmenezes makes all caches timestamp-aware): the fact that the current txn is going to write descriptors and so it needs to make sure it actually reads them first.

We should also assert here that the transaction is SERIALIZABLE and refuse to allow creation of views in SNAPSHOT transactions :S
</blockquote></details>

Aha! This very comment reminds me why I love it when you review a PR. 

I did not really understand what was this thing with "avoiding cached descriptors" before. Now I understand! Thank you!
I added an extensive explanatory comment on the definition of the `avoidCachedDescriptor` field to reflect this newly gained knowledge.

Also I understand that:
- we indeed absolutely need to check that the transaction is SERIALIZABLE for DDL that update descriptors like CREATE VIEW but also anything else that updates descriptors in-place. I'll need to check that, we may have some bugs lurking!
- we are probably not doing the right thing neither for views nor FKs that reference system tables (since the latter create an index and may modify the system descriptors); these txns need to be anchored on the system range with SetSystemConfigTrigger (they are not currently)

This is worth another commit on its own, which I might insert in-between master and this one. I will investigate further. Thank you for the patient explanation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-:-KqXARNh05jle4AQ3bw7:b-gzyjne)*
<!-- Sent from Reviewable.io -->

knz commented:
Filed the two schema problems as separate issues #17404 #17405. They are quite important but can be addressed orthogonally to this PR.

knz committed:
sql: implement new crdb_internal virtual tables to inspect descriptors

The views in pg_catalog and information_schema provide introspection
over the logical schema (pure SQL view) however when testing/debugging
it is also important to see the descriptor IDs and also other fields
of descriptors.

This patch provides help for debugging with the following new virtual
tables:

- `crdb_internal.table_columns`: detailed column IDs and properties
- `crdb_internal.table_indexes`: detailed index IDs and index properties
- `crdb_internal.index_columns`: what columns are indexed and how
- `crdb_internal.backward_dependencies`: which other descriptors each descriptor depends on
- `crdb_internal.forward_dependencies`: which other descriptors each descriptor is depended on by

These virtual tables are intended to display the contents of
descriptors without any data transformation, so as to reveal any error
in the descriptors, if any.

Debugging/fsck tools can later use these to check schema validity.

knz committed:
sql: remove the dependencies column from crdb_internal.create_statements

... because it was incomplete (missing FK relationships).

The two new tables `crdb_internal.forward_dependencies` and
`crdb_internal.backward_dependencies` can be used instead.

knz committed:
sql: re-vamp the view dependency analysis

Prior to this patch, view dependency analysis during CREATE VIEW was
broken because it would only check dependencies after plan
optimization, i.e. possibly after some dependency were lost. For
example, with the queries:

```sql
CREATE VIEW v AS SELECT k FROM (SELECT k,v FROM kv) -- loses dependency on kv.v
CREATE VIEW v AS SELECT k,v FROM kv WHERE FALSE -- loses dependency on kv
```

This patch addresses the issue as follows:

- the dependencies are now collected during the initial construction
  of the query plan, before any optimization are applied. This way we
  ensure the dependency tracking is complete.
- a migration is implemented that will fix any view descriptor and
  corresponding dependency information that was populated prior to
  this fix.

knz commented:
Rebased on top of #17422 to make the fix more obvious.

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 30 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-:-KqdgA2x9hG7h7j6RqW1:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYR

andreimatei commented:
Review status: 0 of 30 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/views.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310#-KqVMwfOgc7OFOEDvlVp:-KuaZRFo7dkgRoR4GEG1:bj910pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9affb59e067cc84b747976da9416cec9692f5e0/pkg/sql/views.go#L79)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Aha! This very comment reminds me why I love it when you review a PR. 

I did not really understand what was this thing with "avoiding cached descriptors" before. Now I understand! Thank you!
I added an extensive explanatory comment on the definition of the `avoidCachedDescriptor` field to reflect this newly gained knowledge.

Also I understand that:
- we indeed absolutely need to check that the transaction is SERIALIZABLE for DDL that update descriptors like CREATE VIEW but also anything else that updates descriptors in-place. I'll need to check that, we may have some bugs lurking!
- we are probably not doing the right thing neither for views nor FKs that reference system tables (since the latter create an index and may modify the system descriptors); these txns need to be anchored on the system range with SetSystemConfigTrigger (they are not currently)

This is worth another commit on its own, which I might insert in-between master and this one. I will investigate further. Thank you for the patient explanation.
</blockquote></details>

This was further discussed in https://github.com/cockroachdb/cockroach/issues/17404

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17310)*
<!-- Sent from Reviewable.io -->
